### PR TITLE
Initial Commit

### PR DESCRIPTION
**Pull request by ig-shaun**
The did:cosmos DID Method specification first draft.
Transferred from https://github.com/EarthProgram/did-cosmos/edit/main/README.md

To display correctly in GitHub pages, this repo would need to be configured for Mermaid, which is used for several of the process diagrams.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,882 @@
...
+
+* **Joe Andrieu** &lt;<a href="mailto:joe@legreq.com">joe@legreq.com</a>&gt;
+* **Lohan Spies** &lt;<a href="mailto:lohan.spies@ixo.world">lohan.spies@ixo.world</a>&gt;
+* **Shaun Conway** &lt;<a href="mailto:shaun.conway@ixo.world">shaun.conway@ixo.world</a>&gt;
+
+This specification is under active development at [https://github.com/interchainio/did-cosmos](https://github.com/interchainio/did-cosmos).
+
+## Abstract
+
+**did:cosmos** is an IID method designed to refer to Cosmos-compatible on-chain assets.
```
Comment (PaddyMc) [inline README.md:15]: Does this mean that we can't use IIDs to reference anything off-chain? That seems to reduce the number of use-cases that this method can satisfy?

What about decentralized identifiers for IOT devices? Or SSI agents? Or chains? Or validators? 

Would this method exclude them?

File: README.md [hunk 2]
```diff
@@ -1 +1,882 @@
...
+
+Restricted to on-chain assets, IIDs are a new class of identifier uniquely suited to the requirements of fungible tokens, non-fungible tokens, and other chain-native components. IID methods can be developed for any compatible blockchain, making them suitable for interoperable representations of assets (and the cryptography that secures those tokens) regardless of the underlying chain.
+
+This particular specification, **did:cosmos** applies the IID approach to assets on blockchains made with the Cosmos SDK.
+
+IIDs also introduce a few new features—conformant extensions to the DID Core specification—that provide for privacy-respecting options for the full range of expected token functionality, including Linked Resources, On-chain Service Endpoints, and Accorded Rights. We describe those here, capturing their definition at the time this specification was developed. See the IID specification for current normative definitions.
```
Comment (PaddyMc) [inline README.md:25]: How are these features' privacy respecting?

File: README.md [hunk 3]
```diff
@@ -1 +1,882 @@
...
+
+DID Methods which conform to the IID specification resolve to a DID document representing how to securely interact with a uniquely identified digital asset, within a unique blockchain namespace. Because IIDs are DIDs, software applications that are conformant with the W3C specification will be able to inter-operate with IIDs and IID documents, although some IID-specific features may require additional tooling.
+
+**did:cosmos** DIDs are IIDs intended identify assets on Cosmos application chains.
```
Comment (PaddyMc) [inline README.md:29]: Aren't most DID use cases a mixture of on and off-chain assets?

For instance in a protocol, there could be:

1. A token => `did:cosmos:token`
2. An agent => `did:cosmos:agent`
3. An IOT device => `did:comsos:IOT`
4. An account managed by a person => `did:cosmos:key:cosmos1asdasdasd`

Here the agent and the IOT device couldn't be represented by this method? Is this correct?

File: README.md [hunk 4]
```diff
@@ -1 +1,882 @@
...
+
+## Architecture Overview
+
+In the Cosmos ecosystem, **application chains** are sovereign blockchains built using the Cosmos SDK. These application chains are comprised of various modules, which perform different functions. We refer to modules that manage the state of on-chain assets as **asset modules**.
+
+To support **did:cosmos**, three modules must be operational: a **chain registry**, a **namespace registry**, and an **asset module**.
```
Comment (PaddyMc) [inline README.md:35]: These are prerequisites to use the DID module that is being proposed?

The seems like cognitive overhead that is use-case specific, no?

File: README.md [hunk 5]
```diff
@@ -1 +1,882 @@
...
+
+**Asset Module** Maintains on-chain assets, with its own module-specific functions for creating, updating, and deactivating assets. Each Asset Module MUST also support a method for retrieving the DID Document for a given asset, given a DID managed by that module.
+
+The role of each of these modules is further described below.
+
+## did:cosmos Syntax
+
+An example `did:cosmos` method identifier is `did:cosmos:version:chainspace:namespace:unique-id`.
+
+There are four forms of did:cosmos DIDs:
```
Comment (PaddyMc) [inline README.md:69]: Is there a way we can have this method split into two cosmos methods? Then have an extension to the spec

1. `did:cosmos:chainspace` 
2. `did:cosmos:chainspace:namespace:unique-id`

File: README.md [hunk 6]
```diff
@@ -1 +1,882 @@
...
+
+```abnf
+did:cosmos:version:chainspace:namespace:unique-id
```
Comment (PaddyMc) [inline README.md:72]: Is the version not managed by DID metadata?

File: README.md [hunk 7]
```diff
@@ -1 +1,882 @@
...
+```
+
+In short, you must have a chainspace, but the version and asset-id are optional.
+
+### did:cosmos ABNF
+
+```abnf
+did-cosmos          = "did:cosmos:" method-specific-id
+method-specific-id = [version ":"] chainspace [":" asset-id]
+version            = 1*DIGIT
```
Comment (PaddyMc) [inline README.md:85]: Is the version not managed by metadata?

File: README.md [hunk 8]
```diff
@@ -1 +1,882 @@
...
+}
+```
+
+## Security Considerations
+
+**did:cosmos** is designed for Cosmos compatible blockchains, which secure chain state using a proof-of-stake algorithm. Actively maintained and supporting hundreds of independent blockchains, Cosmos has a well-developed security process defined at [https://github.com/cosmos/cosmos-sdk/blob/master/SECURITY.md](https://github.com/cosmos/cosmos-sdk/blob/master/SECURITY.md).
+
+All **did:cosmos** DIDs use Secp256k1 keys for initial asset creation and updates. This is the same cryptographic curve used by Bitcoin and Ethereum; it is largely considered secure.
+
+The asset module on application chains is responsible for securing CRUD operations on the assets in manages. A thorough security review is recommended for any asset module.
```
Comment (PaddyMc) [inline README.md:785]: Does this assume that the DID module is always passed into and maintained by an asset module?

**Review by PaddyMc — COMMENTED** [review]
Great work on this method! Left a few questions, mostly just to understand how this would fit into our work! :+1:

File: README.md (hunk 1 above)
Comment (jandrieu) [reply README.md:15]: Not entirely.

The IID way to address off-chain resources is to use an IID Reference or IID Resource, depending on whether you want to refer to the resource (IID Reference) or interact/download the resource (IID Resource).

To wit:
`did:cosmos:ixo:nft:abc123` refers to an on-chain record
`did:cosmos:ixo:nft:abc123#subject` refers to an off-chain resource (quite likely the intended subject of the IID)
`did:cosmos:ixo:nft:abc123/subject` is a URL for interacting with that subject (perhaps a profile page about the subject)

It might help to thing through what is lost when you use DIDs to directly refer to off-chain resources. 

With IIDs, you can construct references to both the on-chain record and any number of off-chain resources
```mermaid
classDiagram
direction LR
IID -->OnChainAsset : refers to
IID Reference -->OffChainResource  : refers to
```
In particular, in html:
```html
<iframe id="a" src="did:cosmos:ixo:nft:abc123"></iframe>
<iframe id="b" src="did:cosmos:ixo:nft:abc123#subject"></iframe>
<iframe id="a" src="did:cosmos:ixo:nft:abc123/subject"></iframe>
```
Each of these three fill differently.
a. will have the DID document
b. will have the DID document, scrolled/focused to the #subject node (possibly in a linkedResource entry)
c. will expose whatever is returned by the endpoint defined in the linkedResource entry

Or consider in RDF (using Turtle)
```turtle
did:cosmos:ixo:nft:abc123 a token:token ;
     a rec:REC2022 ; 
     did:authentication #key1 . 
 
did:cosmos:ixo:nft:abc123#subject a foaf:person;
            foaf:name "Joe Andrieu" . 
did:cosmos:ixo:nft:abc123/subject a social:profile ;
            social:provider https://twitter.com .
```

The first statement states that did:cosmos:ixo:nft:abc123 refers to a token that is an REC2022 renewable energy certificate token, which has the authentication relationship defined by #key1 (defined elsewhere in the DID document)
The second states that did:cosmos:ixo:nft:abc123#subject refers to a person, whose name is "Joe Andrieu".
The third states that did:cosmos:ixo:nft:abc123/subject refers to a social profile, hosted at Twitter.

In contrast, with classic DIDs we have the following 
```mermaid
classDiagram
direction LR
DID1 -->OnChainAsset : refers to
DID2 -->OffChainResource  : refers to
```
Where DID1 = `did:cosmos:ixo:nft:xyz789` and DID2 = `did:cosmos:ixo:nft:abc123`

However, with DIDs there is no way to know--from just the DID itself whether or not one is referring to the on-chain record--which must exist for all blockchain mediated pointers like most DIDs.

That means that in html we have the following use of a two DIDs:
```html
<iframe id="a" src="did:cosmos:ixo:nft:abc123"></iframe>
<iframe id="b" src="did:cosmos:ixo:nft:xyz789"></iframe>
```
Which one should show the DID Document? And which one should show something about the off-chain resource?

This is an unresolvable problem, known as HTTPRange-14 [https://en.wikipedia.org/wiki/HTTPRange-14](https://en.wikipedia.org/wiki/HTTPRange-14)

It's even more clear in the RDF:
```turtle
did:cosmos:ixo:nft:xyz789 a foaf:person;
            foaf:name "Joe Andrieu" . 
did:cosmos:ixo:nft:abc123 a token:token ;
     a rec:REC2022 ; 
     did:authentication #key1 . 
```

The first statement says that `did:cosmos:ixo:nft:xyz789` refers to a person named Joe Andrieu.
The second stays that `did:cosmos:ixo:nft:abc123` refers to a REC2022 token, which has an authentication relationship with the verification method of #key1 (defined somewhere else).

But what is not clear is whether or not `did:cosmos:ixo:nft:abc123` is an off-chain resource that has that authentication relationship merely stated here is an observation, or if it is an on-chain resource in which the authentication relationship is recorded on-chain and reported here as definitive.

The first is a typical canonical use of a DID to refer to a human subject. But how do you refer to the on-chain record that is represented by that DID? You can't. Not with a URL.

The second *could* be considered as pointing to an on-chain record, but there is no way to distinguish that fact. Are statements about `did:cosmos:ixo:nft:abc123` statements about the on-chain asset or are they about some off-chain thing, such as a literal, physical token you might find at an arcade or an amusement park?

It is analogous to the inability to refer to the DNS record of a domain using a URL. There is no transformation of
`http://example.com` that would make the DNS record appear in the following html:
```html
<iframe src="http://example.com"></iframe>
```

TL;DR: if the base identifier refers to the control artifact, aka, the on-chain record that is used to render the DID Document, we can always get to an offchain resource by using a reference fragment (#subject). But the reverse is not true. 

The short answer to the question
> What about decentralized identifiers for IOT devices? Or SSI agents? Or chains? Or validators?
> Would this method exclude them?

Is no, the spec does not prevent using did:cosmos for those subjects. They can use IID References. However, you would NOT use IIDs directly to refer to any of those.

File: README.md (hunk 2 above)
Comment (jandrieu) [reply README.md:25]: They allow the IID definer to choose between the options based on the use case.

In cases where maximum privacy is required, a hashgraph keeps nearly all data off-chain, including even the count of how many resources are associated. 

When manageable privacy is required, a hash with an endpoint allows the endpoint to serve private information using its own business logic, including honoring GDPR deletion requests. Only the hash and the URL of the endpoint are exposed on chain. This, IMO, is the best practice way to identify people associated with an NFT, such as the creator. All personal data is off-chain with privacy mediated by the endpoint service. Putting an artist's name on a blockchain is asking for a GDPR lawsuit.

When privacy (and confidentiality) is known to not be an issue, then linked Resources can simply embed the properties for everyone to see.

Most importantly, IID conformant software will respect all of these variations as equivalents. This means that IID asset definers get to choose what is private and what isn't, and the software just works. It is the same pattern that we see with CSS: inline, linked CSS files, element attributes, and DOM functions can all set styles on HTML elements and all are respected as equally valid. (The only differentiation is the ordering of application of style rules as each rule source gets applied in sequence).

File: README.md (hunk 3 above)
Comment (jandrieu) [reply README.md:29]: > Aren't most DID use cases a mixture of on and off-chain assets?

Exactly right. Which is why IIDs enable you to distinguish between them. Otherwise, we end up conflating the two.

To address these off-chain resources, simply use an IID reference with an IID that refers to the on-chain control record for that resource.

File: README.md (hunk 4 above)
Comment (jandrieu) [reply README.md:35]: I'd say it's less use-case specific than use-case informed.

We developed these requirements through a 6 month+ open working group. The goal was a DID method that could handle all of the NFT use cases people are dealing with.

Yes, many use cases could be addressed with a simpler DID method. `did:key` is one of my favorite methods precisely because for certain use cases, it is brilliant. 

In some ways, the cognitive overhead is similar to CSS. It is a huge conceptual burden to have to learn how to apply this new level of encapsulation and separation of concerns. Many use cases could legitimately be address with plain HTML (no CSS). But that isn't the point. The point is that there are enough high-value use cases to justify the use of CSS. I believe the same is true with IIDs. Baseline DIDs simply do not have a standard solution to many of our requirements. The IID specification is an attempt to work through an approach that **does**.

File: README.md (hunk 5 above)
Comment (jandrieu) [reply README.md:69]: Potentially. @ig-shaun and I talked about some of these options. I'm not sure how the first one in your list refers to anything specific (there's no unique-id), but you might be thinking through some ideas we've been kicking around and presented here. Let me discuss these separately.

The first thing is that you could add an off/on chain flag in the DID itself:
```
did:cosmos:chainspace:namespace:unique-id:onchain
did:cosmos:chainspace:namespace:unique-id:offchain
```

And, potentially, you could suppose that the default is one or the other, say `onchain`. Then for on-chain resources the DID would look the same, but offchain resources would have "offchain" in the URI. This ends up being syntactically almost the same as using IID References (and you could decide at the method specification level which one is default). However, whichever is default, you would lose the ability to have multiple references per IID. For example, in a given NFT, you might need to refer to half a dozen real-world resources, all of which can have their own IID Reference within the NFT's IID namespace. For example, a reference to an address, a parcel, a jurisdiction, and a zip code all as part of a real property title. IIDs can handle that, DIDs alone not so much.

The second thing is the innovation in our syntax that looks a lot like the first entry you listed `did:cosmos:chainspace`. This is essentially just syntactic shorthand for referring to the chainspace record on the chain registry even when you don't know the IID of that record. This allows one to strip off the first part of the DID to go get the DID Document for the chainspace itself. That DID Document would probably have a schema that maps to an easy download of the chain.json (`did:cosmos:chainspace/chain.json`) or provides the service endpoints for updating the chain registry entry (`did:cosmos:chainspace/update`).

However, since that first item doesn't specify an individual asset (other than the chainspace), I might have misunderstood your question.

File: README.md (hunk 6 above)
Comment (jandrieu) [reply README.md:72]: If the version change is a breaking change to how your resolve the DID itself, then you don't have a way to get that DID metadata.

My initial preference was to support a version 0 that explicitly says that resolution starts by looking up the entry in the chain registry github repo, and a later version 1 that would start by querying the on-chain chain registry. However, it is inevitable that the repo and the on-chain registry would, eventually, get out of sync. In that case, which is authoritative?

And more importantly, if old software was built for version 0, it would have no way to know that its algorithm is out of date. It would have no way to know that for this new DID, you should FIRST check the on-chain registry and should not use the github repo. However, with a version in the DID itself, the software can simply verify that the DID it is processing uses a version it understands. Of course, you can't magically get past modifying code to support the new feature, but you can avoid attempting a knowably out-of-date resolution process.

It's these sorts of breaking changes--when extant DIDs in the wild are no longer compatible with the current version--that are addressed by explicitly naming the version in the DID itself.

File: README.md (hunk 7 above)
Comment (jandrieu) [reply README.md:85]: See previous comment

File: README.md (hunk 8 above)
Comment (jandrieu) [reply README.md:785]: Maybe.

I'm not sure if I fully understand what you mean by passing a DID module into an asset module.

The IID architecture assumes you already have some on-chain asset management module. For example:
* An account module. 
* An NFT module. 
* A KYC module. 
* Whatever. 

Those asset modules ALREADY define messages for triggering functions on those modules. IIDs don't change that. The module keeps doing what it does in the same way its always done it (and they can continue to innovate and add new functionality). 

What each asset module DOES need to do is to render a DID Document that represents the verification relationships, verification methods, and services for a given asset as defined by a unique ID. 

Fortunately, each individual asset module is in the best position to decide exactly what should go in such a DID Document and it need only render that at the point of inquiry--it need not store anything additional on chain. It already is storing enough on-chain to identify and secure its assets, so it just needs to map that into a DID Document and return it.

**Review by jandrieu — COMMENTED** [review]
@PaddyMc Thanks for the detailed questions. I tried to answer them each in comments. Let me know if you still have questions. 

If there are edits to the spec that you think would help others with similar questions, we'd welcome those as well.

**Comment by jandrieu** [comment]
@PaddyMc Btw, sorry those mermaid diagrams don't render in the comment flow. They worked great in the comment preview. =(

File: README.md (hunk 1 above)
Comment (PaddyMc) [reply README.md:15]: Great, thanks for the detailed explanation!

File: README.md (hunk 2 above)
Comment (PaddyMc) [reply README.md:25]: Nice, so the Hashgraph enables privacy. To use this privacy-respecting feature an IID definer would add a `linkedResource` to their DID of type `hashgraph` with a proof and a service endpoint to enable some communication about the Hashgraph?

File: README.md (hunk 4 above)
Comment (PaddyMc) [reply README.md:35]: Ok, I understand! Could this be applied to blockchains leveraging the IBC protocol, instead of modules that live on the same blockchain?

File: README.md (hunk 5 above)
Comment (PaddyMc) [reply README.md:69]: Thanks for clearing this up! The question was more about how to ease the burden of understanding for new implementers of the standard, but good info all the same :+1:

File: README.md (hunk 6 above)
Comment (PaddyMc) [reply README.md:72]: hmmm, I understand the reasoning here but I think in the wild there would be ways to mitigate this such as data migration code and correct proto versioning, but defo best to keep it until we can fully understand the technical implementation details.

File: README.md (hunk 8 above)
Comment (PaddyMc) [reply README.md:785]: Dependency inject the did module into the asset module so the asset module can CRUD on the did module state. 

So an asset module will manage the lifecycle of an IID document and the IID module will not be exposed to an IID definer?

**Comment by PaddyMc** [comment]
Thanks so much for the detailed responses @jandrieu ! And apologies for the delayed response, been a hectic few weeks!

Last question:

Is there some sort of reference implementation somewhere? Or what's the status of that? :pray:
